docs/design.md

```diff
@@ -0,0 +1,481 @@
+
+Merkle trees provide the memory commitment mechanism due to their:
+
+- Challenge-independent commitment through initial and final root hashes
+- Natural sparse memory handling via partial tree pruning of unused intermediate
+  nodes
+
+Efficient recursion requires a ZK-friendly hash function. The current
+implementation uses Poseidon2, though alternative hash functions can be
+substituted as needed.
+
+The memory address space spans `0..2**30`, eliminating Merkle root padding
+requirements while providing sufficient capacity for substantial computations.
+Extension to `P - 1` is possible but unnecessary for client-side proving
+scenarios, where memory requirements remain modest compared to long-trace
+applications. Memory exceeding 2^30 is only required for extremely long runs,
+which would demand RAM capacity beyond typical consumer device specifications.
+
+The Merkle tree omits leaf hashing to minimize computational overhead. Although
+leaf hashing prevents sibling value disclosure in inclusion proofs, it provides
+no benefit for state root proving and is therefore omitted. This approach
+requires all memory cells to contain values, resulting in implicit
+zero-initialization of the entire memory segment. The default zero value has no
+practical impact since the VM overwrites cells with actual values during
+execution.
+
+### Read/Write operations
+
+To support arbitrarily long programs within a fixed-size memory segment (2^30 =
+1,073,741,824 addresses), the design employs a read-write memory model.
+
+Read and write operations are implemented through lookup arguments:
+
+#### Memory Model
+
+- **Memory entries**: triplets `(address, clock, value)` where `clock`
+  timestamps when `address` contained `value`
+- **Clock sequence**: monotonic counter from 0, determined during witness
+  generation
+- **Initial values**: emitted from Merkle commitment with `clock = 0`
+
+#### Operation Mechanics
+
+- **Read/write unification**: Both operations cancel
+  `-(address, prev_clock, prev_value)` and add `+(address, clock, value)` to the
+  lookup sum
+- **Read optimization**: When `value == prev_value`, duplicate storage is
+  avoided and only one column is required
+- **Clock monotonicity**: Enforced via `RangeCheck` component
+  (`clock - prev_clock > 0`)
+
+#### Clock Update Component
+
+- **Purpose**: Bridges large temporal gaps between memory operations
+- **Trigger**: When clock difference exceeds `RangeCheck` capacity
+- **Implementation**: Updates clock values similarly to read operations
+- **Generation**: Trace filled during witness creation, outside VM execution
+- **Strategy**: Divides large deltas into range-check-compliant segments
+
+#### Column Cost Analysis
+
+**Read-write memory** (per access):
+
+- Main trace: 5 columns (`address`, `prev_clock`, `clock`, `prev_value`,
+  `value`)
+- Lookup: 3 columns (subtract previous, add current, range check)
+- Total: 5T + 3L
```

Comment:
T and L are never explicitly defined.
Also introducing the section with why they are different and what is QM31 could help with the context?